docs/guide/counter/plugins/counter/counter.go

```diff
@@ -1,98 +1,197 @@
+	ctr, err := checkTx(ctx, tx)
+	if err != nil {
+		return res, err
 	}
-	if !tx.Fee.IsValid() {
-		return abci.ErrInternalError.AppendLog("CounterTx.Fee is not sorted or has zero amounts")
+	// note that we don't assert this on CheckTx (ValidateBasic),
+	// as we allow them to be writen to the chain
+	if !ctr.Valid {
+		return res, ErrInvalidCounter()
 	}
-	if !tx.Fee.IsNonnegative() {
-		return abci.ErrInternalError.AppendLog("CounterTx.Fee must be nonnegative")
+
+	// handle coin movement.... like, actually decrement the other account
+	if !ctr.Fee.IsZero() {
+		// take the coins and put them in out account!
+		senders := ctx.GetPermissions("", stack.NameSigs)
+		if len(senders) == 0 {
+			return res, errors.ErrMissingSignature()
+		}
+		in := []coin.TxInput{{Address: senders[0], Coins: ctr.Fee, Sequence: ctr.Sequence}}
+		out := []coin.TxOutput{{Address: CounterAcct(), Coins: ctr.Fee}}
+		send := coin.NewSendTx(in, out)
+		// if the deduction fails (too high), abort the command
+		_, err = dispatch.DeliverTx(ctx, store, send)
+		if err != nil {
+			return res, err
+		}
 	}
 
-	// Did the caller provide enough coins?
-	if !ctx.Coins.IsGTE(tx.Fee) {
-		return abci.ErrInsufficientFunds.AppendLog("CounterTx.Fee was not provided")
+	// update the counter
+	state, err := LoadState(store)
+	if err != nil {
+		return res, err
 	}
+	state.Counter += 1
+	state.TotalFees = state.TotalFees.Plus(ctr.Fee)
+	err = StoreState(store, state)
 
-	// TODO If there are any funds left over, return funds.
-	// e.g. !ctx.Coins.Minus(tx.Fee).IsZero()
-	// ctx.CallerAccount is synced w/ store, so just modify that and store it.
+	return res, err
+}
 
-	// Load CounterPluginState
-	var cpState CounterPluginState
-	cpStateBytes := store.Get(cp.StateKey())
-	if len(cpStateBytes) > 0 {
-		err = wire.ReadBinaryBytes(cpStateBytes, &cpState)
-		if err != nil {
-			return abci.ErrInternalError.AppendLog("Error decoding state: " + err.Error())
-		}
+func checkTx(ctx basecoin.Context, tx basecoin.Tx) (ctr CounterTx, err error) {
+	ctr, ok := tx.Unwrap().(CounterTx)
+	if !ok {
+		return ctr, errors.ErrInvalidFormat(tx)
+	}
+	err = ctr.ValidateBasic()
+	if err != nil {
+		return ctr, err
 	}
+	return ctr, nil
+}
 
-	// Update CounterPluginState
-	cpState.Counter += 1
-	cpState.TotalFees = cpState.TotalFees.Plus(tx.Fee)
+// CounterStore
+//--------------------------------------------------------------------------------
 
-	// Save CounterPluginState
-	store.Set(cp.StateKey(), wire.BinaryBytes(cpState))
+func CounterAcct() basecoin.Actor {
+	return basecoin.Actor{App: NameCounter, Address: []byte{0x04, 0x20}}
+}
 
-	return abci.OK
+type CounterState struct {
+	Counter   int         `json:"counter"`
+	TotalFees types.Coins `json:"total_fees"`
 }
 
-func (cp *CounterPlugin) InitChain(store types.KVStore, vals []*abci.Validator) {
+func StateKey() []byte {
+	return []byte(NameCounter + "/state")
```

Comment:
Is the state space not dictated by basecoin? does each plugin get to choose its own state space like here? This could be a problem for plugins hacking plugins, not sure this should exist on the plugin side as it does here, but maybe I'm just confused